synse_server/plugin.py

```diff
@@ -196,33 +238,47 @@ def refresh(self) -> None:
             logger.debug('refreshing plugin manager')
             start = time.time()
 
-            for address, protocol in self.load():
+            plugins = []
+            plugins.extend(self.load())
+            plugins.extend(self.discover())
+
+            existing, new, removed = self.bucket_plugins(plugins)
+            logger.debug('bucketed plugins', existing=existing, new=new, removed=removed)
+
+            # Register all new plugins
+            for plugin in new:
                 try:
-                    self.register(address=address, protocol=protocol)
+                    self.register(address=plugin[0], protocol=plugin[1])
                 except Exception as e:
                     # Do not raise. This could happen if we can't communicate with
                     # the configured plugin. Future refreshes will attempt to re-register
                     # in this case. Log the failure and continue trying to register
                     # any remaining plugins.
                     logger.warning(
                         'failed to register configured plugin - will attempt re-registering later',
-                        address=address, protocol=protocol, error=e,
+                        address=plugin[0], protocol=plugin[1], error=e,
                     )
                     continue
 
-            for address, protocol in self.discover():
-                try:
-                    self.register(address=address, protocol=protocol)
-                except Exception as e:
-                    # Do not raise. This could happen if we can't communicate with
-                    # the configured plugin. Future refreshes will attempt to re-register
-                    # in this case. Log the failure and continue trying to register
-                    # any remaining plugins.
-                    logger.warning(
-                        'failed to register discovered plugin - will attempt re-registering later',
-                        address=address, protocol=protocol, error=e,
+            # Disable all removed plugins and stop any active tasks they may be running.
+            for plugin in removed:
+                logger.info(
+                    'registered plugin not found during refresh, marking as disabled',
+                    plugin=plugin,
+                )
+                plugin.disabled = True
+                plugin.cancel_tasks()
+
+            # Check if the existing plugin was disabled. If so, re-enable it. Otherwise, there
+            # is nothing to do here.
+            for plugin in existing:
+                if plugin.disabled:
+                    logger.info(
```

Comment:
+1 (disable should be disabled in the log statement)

Reply:
good catch - thanks!